src/Components/Web.Extensions/src/InputFile/RemoteBrowserFileStream.cs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Web.Extensions
+{
+    internal class RemoteBrowserFileStream : BrowserFileStream
+    {
+        private readonly IJSRuntime _jsRuntime;
+        private readonly ElementReference _inputFileElement;
+        private readonly int _maxChunkSize;
+        private readonly PipeReader _pipeReader;
+        private readonly CancellationTokenSource _fillBufferCts;
+
+        private bool _isReadingCompleted;
+        private bool _isDisposed;
+
+        public RemoteBrowserFileStream(IJSRuntime jsRuntime, ElementReference inputFileElement, int maxChunkSize, int maxBufferSize, BrowserFile file)
+            : base(file)
+        {
+            _jsRuntime = jsRuntime;
+            _inputFileElement = inputFileElement;
+            _maxChunkSize = maxChunkSize;
+
+            var pipe = new Pipe(new PipeOptions(pauseWriterThreshold: maxBufferSize, resumeWriterThreshold: maxBufferSize));
+            _pipeReader = pipe.Reader;
+            _fillBufferCts = new CancellationTokenSource();
+
+            _ = FillBuffer(pipe.Writer, _fillBufferCts.Token);
+        }
+
+        private async Task FillBuffer(PipeWriter writer, CancellationToken cancellationToken)
+        {
+            long offset = 0;
+
+            while (offset < File.Size)
+            {
+                var pipeBuffer = writer.GetMemory(_maxChunkSize);
+                var chunkSize = (int)Math.Min(_maxChunkSize, File.Size - offset);
+
+                try
+                {
+                    var base64 = await _jsRuntime.InvokeAsync<string>(
```

Comment:
I think System.Text.Json will decode base64 encoded text. The other option would be to write a custom type with a JsonConverter that can deserialize a JSON array for us. That would avoid this two step conversion.

Reply:
Let's not tackle the Base64 conversion right now. We could do a quick try deserializing it as a byte array, but let's not go further than that